### PR TITLE
Feature configparse redux

### PR DESCRIPTION
**Pull request by northwestwitch**
This PR contains the same code as #2496, which I've merged to master by mistake (and then reverted) today. Sorry @mikaell !!  Fix #2496

**How to test**:
1. how to test it, possibly with real cases/data

**Expected outcome**:
The functionality should be working
Take a screenshot and attach or copy/paste the output.

**Review:**
- [x] code approved by CR
- [x] tests executed by CR, ML

### CODE REVIEW

**Comment by northwestwitch** [comment]
Time to test this PR with real data.

1) Installed on hasta stage:
image

1) First test, load a duplicate of [this cancer case](https://scout-stage.scilifelab.se/cust000/actualgecko)
I get an error:
image

**Comment by mikaell** [comment]
I'll have a look!

**Comment by northwestwitch** [comment]
I can reproduce the error locally with the demo case if I remove the capture-kit - related lines in the demo file:

```
Traceback (most recent call last):
  File "/Users/chiararasi/miniconda3/envs/py38/bin/scout", line 33, in <module>
    sys.exit(load_entry_point('scout-browser', 'console_scripts', 'scout')())
  File "/Users/chiararasi/miniconda3/envs/py38/lib/python3.8/site-packages/click/core.py", line 1137, in __call__
    return self.main(*args, **kwargs)
  File "/Users/chiararasi/miniconda3/envs/py38/lib/python3.8/site-packages/flask/cli.py", line 596, in main
    return super().main(*args, **kwargs)
  File "/Users/chiararasi/miniconda3/envs/py38/lib/python3.8/site-packages/click/core.py", line 1062, in main
    rv = self.invoke(ctx)
  File "/Users/chiararasi/miniconda3/envs/py38/lib/python3.8/site-packages/click/core.py", line 1668, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/Users/chiararasi/miniconda3/envs/py38/lib/python3.8/site-packages/click/core.py", line 1668, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/Users/chiararasi/miniconda3/envs/py38/lib/python3.8/site-packages/click/core.py", line 1404, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/Users/chiararasi/miniconda3/envs/py38/lib/python3.8/site-packages/click/core.py", line 763, in invoke
    return __callback(*args, **kwargs)
  File "/Users/chiararasi/miniconda3/envs/py38/lib/python3.8/site-packages/click/decorators.py", line 26, in new_func
    return f(get_current_context(), *args, **kwargs)
  File "/Users/chiararasi/Documents/work/GITs/scout/scout/commands/setup/setup_scout.py", line 192, in demo
    setup_scout(
  File "/Users/chiararasi/Documents/work/GITs/scout/scout/load/setup.py", line 202, in setup_scout
    config_data = parse_case_data(config=case_data)
  File "/Users/chiararasi/Documents/work/GITs/scout/scout/parse/case.py", line 50, in parse_case_data
    config_dict = parse_case_config(config)
  File "/Users/chiararasi/Documents/work/GITs/scout/scout/parse/case.py", line 111, in parse_case_config
    parsed_config = ScoutLoadConfig(**config)
  File "/Users/chiararasi/Documents/work/GITs/scout/scout/parse/models.py", line 191, in __init__
    super().__init__(vcf_files=vcfs, **data)
  File "pydantic/main.py", line 406, in pydantic.main.BaseModel.__init__
pydantic.error_wrappers.ValidationError: 3 validation errors for ScoutLoadConfig
samples -> 0 -> capture_kit
  field required (type=value_error.missing)
samples -> 1 -> capture_kit
  field required (type=value_error.missing)
samples -> 2 -> capture_kit
  field required (type=value_error.missing)
```

**Comment by mikaell** [comment]
I found the bug. Will do some more testing.

**Comment by northwestwitch** [comment]
> When testing my patch on Hasta I run into problems -any ideas?
> 
> ```
> [0|0|80] [mikael.laaksonen@hasta:/home/proj/stage/housekeeper-bundles/safeguinea/2021-05-30] [S_scout] $ scout load case scout_load.yaml
> 2021-09-29 16:00:56 hasta.scilifelab.se scout.commands.base[196191] INFO Running scout version 4.39
> 2021-09-29 16:00:56 hasta.scilifelab.se scout.commands.base[196191] DEBUG Debug logging enabled.
> 2021-09-29 16:00:56 hasta.scilifelab.se scout.parse.case[196191] DEBUG Checking for SMN TSV..
> 2021-09-29 16:00:56 hasta.scilifelab.se scout.parse.case[196191] INFO Adding SMN info from /home/proj/stage/housekeeper-bundles/safeguinea/2021-05-30/safeguinea_smn.tsv.
> 2021-09-29 16:00:56 hasta.scilifelab.se scout.parse.case[196191] DEBUG Checking for SMN TSV..
> 2021-09-29 16:00:56 hasta.scilifelab.se scout.parse.case[196191] INFO Adding SMN info from /home/proj/stage/housekeeper-bundles/safeguinea/2021-05-30/safeguinea_smn.tsv.
> 2021-09-29 16:00:56 hasta.scilifelab.se scout.commands.load.case[196191] INFO Use family safeguinea
> 2021-09-29 16:00:56 hasta.scilifelab.se scout.adapter.mongo.institute[196191] DEBUG Fetch institute cust000
> 2021-09-29 16:00:57 hasta.scilifelab.se scout.commands.load.case[196191] ERROR Unhandled Exception: Traceback (most recent call last):
>   File "/home/proj/bin/conda/envs/S_scout/lib/python3.7/site-packages/scout/commands/load/case.py", line 122, in case
>     adapter.load_case(config_data, update, keep_actions)
>   File "/home/proj/bin/conda/envs/S_scout/lib/python3.7/site-packages/scout/adapter/mongo/case.py", line 570, in load_case
>     institute_obj = self.institute(config_data["owner"])
>   File "/home/proj/bin/conda/envs/S_scout/lib/python3.7/site-packages/scout/adapter/mongo/institute.py", line 189, in institute
>     institute_obj = self.institute_collection.find_one({"_id": institute_id})
>   File "/home/proj/bin/conda/envs/S_scout/lib/python3.7/site-packages/pymongo/collection.py", line 1319, in find_one
>     for result in cursor.limit(-1):
>   File "/home/proj/bin/conda/envs/S_scout/lib/python3.7/site-packages/pymongo/cursor.py", line 1207, in next
>     if len(self.__data) or self._refresh():
>   File "/home/proj/bin/conda/envs/S_scout/lib/python3.7/site-packages/pymongo/cursor.py", line 1100, in _refresh
>     self.__session = self.__collection.database.client._ensure_session()
>   File "/home/proj/bin/conda/envs/S_scout/lib/python3.7/site-packages/pymongo/mongo_client.py", line 1816, in _ensure_session
>     return self.__start_session(True, causal_consistency=False)
>   File "/home/proj/bin/conda/envs/S_scout/lib/python3.7/site-packages/pymongo/mongo_client.py", line 1766, in __start_session
>     server_session = self._get_server_session()
>   File "/home/proj/bin/conda/envs/S_scout/lib/python3.7/site-packages/pymongo/mongo_client.py", line 1802, in _get_server_session
>     return self._topology.get_server_session()
>   File "/home/proj/bin/conda/envs/S_scout/lib/python3.7/site-packages/pymongo/topology.py", line 488, in get_server_session
>     None)
>   File "/home/proj/bin/conda/envs/S_scout/lib/python3.7/site-packages/pymongo/topology.py", line 217, in _select_servers_loop
>     (self._error_message(selector), timeout, self.description))
> pymongo.errors.ServerSelectionTimeoutError: localhost:27017: [Errno 111] Connection refused, Timeout: 0.02s, Topology Description: <TopologyDescription id: 615471984517c45bf943ae9a, topology_type: Single, servers: [<ServerDescription ('localhost', 27017) server_type: Unknown, rtt: None, error=AutoReconnect('localhost:27017: [Errno 111] Connection refused')>]>
> 
> Aborted!
> ```

Are you running your scripts from the stage environment (`us` command?)

**Comment by mikaell** [comment]
Errors reported by @northwestwitch (scout load case scout_load.yaml) are now fixed.

**Comment by mikaell** [comment]
It seems like all tests pass. Ready for one more round of reviews!

**Comment by dnil** [comment]
Ok, db has moved so we should try to get this in soonish. No stress, but let us know if/how we can help best - except for doing what we can to avoid changing parsing of anything on master. 😬

**Comment by mikaell** [comment]
Hi!
This is read for a review and run through to find the bugs I've missed.

File: scout/parse/models.py [hunk 1]
```diff
@@ -0,0 +1,419 @@
...
+        return self.title == other.title
+
+    @root_validator
+    def valid_image_suffix(cls, values):
+        path = values.get("path")
+        path = Path(path)
+
+        # Skip configured image if path suffix is not an image file type
+        if not path.suffix[1:] in ["gif", "svg", "png", "jpg", "jpeg"]:
+            values["path"] = None
```
Comment (northwestwitch) [inline scout/parse/models.py:88]: Looks like the idea is skipping the load of a custom image if the format is not standard (["gif", "svg", "png", "jpg", "jpeg"]). But what happens if you try to load a wrong image format instead ,is that the load will crash with this error:

```
2021-11-15 10:36:45 ChiaraRMBP.local scout.parse.models[55318] DEBUG IMAGE:data=None description='A very good description' height=100 format=None path=None str_repid='AFF2' title='A png image' width=500
Traceback (most recent call last):
  File "/Users/chiararasi/miniconda3/envs/py38/bin/scout", line 33, in <module>
    sys.exit(load_entry_point('scout-browser', 'console_scripts', 'scout')())
  File "/Users/chiararasi/miniconda3/envs/py38/lib/python3.8/site-packages/click/core.py", line 1128, in __call__
    return self.main(*args, **kwargs)
  File "/Users/chiararasi/miniconda3/envs/py38/lib/python3.8/site-packages/flask/cli.py", line 600, in main
    return super().main(*args, **kwargs)
  File "/Users/chiararasi/miniconda3/envs/py38/lib/python3.8/site-packages/click/core.py", line 1053, in main
    rv = self.invoke(ctx)
  File "/Users/chiararasi/miniconda3/envs/py38/lib/python3.8/site-packages/click/core.py", line 1659, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/Users/chiararasi/miniconda3/envs/py38/lib/python3.8/site-packages/click/core.py", line 1659, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/Users/chiararasi/miniconda3/envs/py38/lib/python3.8/site-packages/click/core.py", line 1395, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/Users/chiararasi/miniconda3/envs/py38/lib/python3.8/site-packages/click/core.py", line 754, in invoke
    return __callback(*args, **kwargs)
  File "/Users/chiararasi/miniconda3/envs/py38/lib/python3.8/site-packages/click/decorators.py", line 26, in new_func
    return f(get_current_context(), *args, **kwargs)
  File "/Users/chiararasi/Documents/work/GITs/scout/scout/commands/setup/setup_scout.py", line 192, in demo
    setup_scout(
  File "/Users/chiararasi/Documents/work/GITs/scout/scout/load/setup.py", line 202, in setup_scout
    config_data = parse_case_data(config=case_data)
  File "/Users/chiararasi/Documents/work/GITs/scout/scout/parse/case.py", line 52, in parse_case_data
    config_dict = parse_case_config(config)
  File "/Users/chiararasi/Documents/work/GITs/scout/scout/parse/case.py", line 114, in parse_case_config
    parsed_config = ScoutLoadConfig(**config)
  File "/Users/chiararasi/Documents/work/GITs/scout/scout/parse/models.py", line 312, in __init__
    super().__init__(vcf_files=vcfs, **data)
  File "pydantic/main.py", line 406, in pydantic.main.BaseModel.__init__
pydantic.error_wrappers.ValidationError: 3 validation errors for ScoutLoadConfig
custom_images -> __root__
  argument of type 'NoneType' is not iterable (type=type_error)
custom_images -> __root__
  expected str, bytes or os.PathLike object, not NoneType (type=type_error)
custom_images -> __root__
  expected str, bytes or os.PathLike object, not NoneType (type=type_error)
```

Which I think is also why the test `tests/parse/test_parse_case.py::test_parse_incorrect_custom_images` is failing at the moment.

**Comment by mikaell** [comment]
This is ready for a re-review! (and also some external energy to keep me going). 

It should pass all tests.

**Comment by dnil** [comment]
Superb timing - we were just about to change the config parsing!

**Comment by dnil** [comment]
Well, not case config parsing luckily. 😁

**Comment by northwestwitch** [comment]
I think this is fine to merge now. We did test it by loading both a RD and cancer case immediately before that change that allows to image files for variants and it worked fine. I'd say let's merge!

**Pull request merged**